### ISSUE TITLE
A REST interface for NetworkAPI

### ISSUE THREAD

**Issue by dkeeney** [htm-community/htm.core#773]
With the latest discussions on the forum about a REST interface to htm.core I thought I would take a look at what it would take to build it.  It seems rather simple...

The first step would be to add a configure( ) function to the Network class that accepted a YAML string that could be used to configure everything.     
```
   * Syntax:
   *      network:
   *         - registerRegion:   (one of these for each non built-in region)
   *               type: <region type>
   *               path: <path to shared lib to link to>
   *               class: <classname to load>
   *
   *        -  addRegion:  (one of these for each region to be configured in execution order)
   *               name: <region name>
   *               type: <region type>
   *               params: <list of parameters>
   *               phase:  <optonal phase number>
   *
   *         - addLink:  (one of these for each link to be connected)
   *               srcName: <Name of the source region>
   *               destName: <Name of the destination region>
   *               srcOutput: <Name of the source output>
   *               destInput: <Name of the destination input>
   *               propagationDelay: <iterations to delay>
```

The 'registerRegion' would need to be able to register a Python class or a C++ class in a shared library that can be loaded at runtime.

Then I would need a network server loop that would watch the LAN for an incoming message.  If it was a configure message, call the configure( ) function.  It it was data, deserialize it, pass it to a specified encoder and execute run( 1 ) to process it.  Then return a serialized result from a specified output.
```
Syntax:
      msg: 
          dest:  <destination region . Input>
          src:     <returned region . output>
          type:  <data type>
          data:  <array of serialized data>

      response:
           type:  <data type>
           data:  <array of serialized data >

      error: <error message from exception>
```

**Comment by breznak** [comment]
I'd love this! 

I'll have to look at it deeper later but the idea is great! 
Would this go with an additional constructor that accepts JSON? 
```
TM(numCols, activationThreshold, ...)

TM(JSON config)
```
That would be useful for param optimization and API compatibility imho.

**Comment by dkeeney** [comment]
> Would this go with an additional constructor that accepts JSON?

yes, the same parser can handle both YAML and JSON.

A key part to this is the ability to load either a Python or C++ plugin at runtime.  So that may be the first step in completing the REST interface.

**Comment by dkeeney** [comment]
My initial investigation of this project I found that doing the "RegisterRegion" operation to load a non built-in C++ plugin is going to take quite a significant amount of work.  Doable, but it would delay doing the rest of the REST interface.  So, what I thought I would do is set that part aside for now.  We would still be able to use all of the built-in plugins and any Python plugin that you might want to register.  Someone using this would just not be able to construct their own C++ plugin without compiling it into the htm.core library.  They could still load user-written Python plugins.

My thinking for the REST feature is a new "RestAPI" class that resides in src/htm/engine/restAPI.hpp and .cpp.  This would contain the generic code to implement the REST interface and then would call into the library to perform operations.

There will be a separate 'server.cpp' which is a 'main' that does nothing but call an entry point in the RestAPI class.  Anyone using this library could embed the RestAPI into their own executable or they could use our 'main' for a stand-alone server executable.

I will need a new 'configure( )' function added to the Network class to perform the configuration step; parsing the YAML or JSON string and adding and configuring all of the regions and data links between them.  

I will need some new logic that accepts data from the REST api and passes it to the input of the first region in the chain setup with the configuration( ).  Then accept the output from the last region and return it as the results of the REST call.

Then I need a Unit test for all of this.  Perhaps one thread to run the server part and a second thread to pass data to it and inspect the results to make sure it returns what is expected.

Some documentation would be nice too.

And that is about all it takes....

**Comment by dkeeney** [comment]
I wanted to use an off-the-shelf REST server library.  The two really good ones are [boost::Beast](https://www.boost.org/doc/libs/1_72_0/libs/beast/doc/html/index.html) and [cpprestsdk](https://github.com/Microsoft/cpprestsdk) both use boost.  

There is nothing wrong with the Boost library but it takes so long to download and is huge.  We don't want to include Boost inside htm.core.  But what we can do is have the stand-alone main use Beast or cpprestsdk along with the htm.core library to build the REST server.  The htm.core library just provides handlers for 'configure' and 'data' request operations.  

That allows users to use whatever REST package they would like (and I don't have to write it).  

- User downloads and installs Boost
- User downloads htm.core
- User builds server.exe from included sample code to create the REST api server for htm.core.

The new RESTapi class would contain those two handlers and perhaps any other handlers that we find we need.

**Comment by Thanh-Binh** [comment]
sorry for my long inactive time. For now, our htm.core includes many new extensions, so that I really do not know if everything is controllable. I have a feeling that we should maintain the main C++ core (or I say the lowest level in htm/algorithms, NetworkAPI) as minimum as possible, after compiling we have a core library like "libhtmcore.so". Others for better configuration /easy-to-use interfaces or for support other languages like Python, C# etc will be defined as extension.
From the points of views of users, the users can choose different compiler options as they want, and for some embedded applications, they need only very very small pure C++ core library "libhtmcore.so".
What do you think? Thanks

**Comment by dkeeney** [comment]
Not sure I follow what you are saying.
The lowest level are the htm algorithms.
A wrapper on those is NetworkAPI
The REST interface I am proposing is just an alternative way to invoke NetworkAPI.

The htm algorithms and NetworkAPI components are in the htm.core library and I agree we should keep that small as possible.

What I am proposing for the REST api is to make a few minor additions to NetworkAPI so that an external REST server could access the NetworkAPI methods to to handle requests.

1) a single configure( ) function that can perform all configuration with one YAML or JSON string.
2) A way to pass in data for NetworkAPI to use.  (pass data to encoder from network request)
3) A way to return data from NetworkAPI. (generate response from a region's output)

I will be including this external REST server code as an example.   The server is the part that handles all of the web message protocols and network I/O.  It will not be included in the htm.core library.

**Comment by dkeeney** [comment]
Lets take a look at the HelloSPTP example.  This is an example implemented using only low level algorithm calls.  It configures a sequence of an encoder, an SP, and a TM.  The data for this example is a set of points on a sine wave.  After a few iterations it tries to predict the likelyhood of the next point on the sine wave.

The examle napi_sine tries to do the same thing using the NetworkAPI.  It configures the encoder, SP, and TP.   Then it iterates these with data from a sine wave (0.01 radians per iteration).

What I want to do with the REST api is to use one REST message call to create an instance of the Network object and configure it with the three regions.  Then subsequent REST message calls would pass in data for one iteration, passing it to the encoder and retrieving data from the TM and returning it.  So rather than being in a loop, the Network api would perform one iteration each time there is a REST request message.

**Comment by dkeeney** [comment]
I should point out that the predicted output of the TM is not a point on the sine wave but an SDR that contains the semantics of that point.  To get back to a real predicted point on the sine wave we would also need to call a classifier that has been trained on mapping the result SDR to the physical point.

So to be useful to the REST api in this case we might want to add a ClassifierRegion component to the configuration.  The returned response to a request would be the highest probability output of the Classifier.  It all depends on what the REST application is trying to do.

**Comment by dkeeney** [comment]
Lets look at how we might configure the Network class.  

```
    ///////////////////////////////////////////////////////////////
    //
    //                          .----------------.
    //                         |    encoder      |
    //         request data--->|  (RDSERegion)   |
    //                         |                 |
    //                         `-----------------'
    //                                 |
    //                         .-----------------.
    //                         |   sp_global     |
    //                         |  (SPRegion)     |
    //                         |                 |
    //                         `-----------------'
    //                                 |
    //                         .-----------------.
    //                         |      tm         |
    //                         |   (TMRegion)    |
    //                         |                 |
    //                         `-----------------'
    //                                 |
    //                         .-----------------.
    //                         |    classifier   |
    //                         |(ClassifierRegion)|
    //                         |                 | ---> response data
    //                         `-----------------'
    //
    //////////////////////////////////////////////////////////////////
```
Note that I added a Classifier region to the napi_sine example.  This is so we can pass back a real point on the sine wave rather than a serialized SDR. 

Rather than performing individual calls to addRegion( ) and Link( ), lets put this in a text string so it can be passed in one REST request to setup the run.  In JSON syntax it might look like this:
```
{network: {
   addRegion: [
      {name: encoder, type: RDSERegion, params: {sparsity: 0.2, size: 1000, radius: 0.03, seed: 2019, noise: 0.01}},
      {name: sp_global, type: SPRegion, params: {columnCount: 2048}},
      {name: tm, type: TMRegion, params: {cellsPerColumn: 8, orColumnOutputs: true}},
      {name: classifier, type: ClassifierRegion, params: {learn: true}}],
    addLink: [
     {src: encoder.encoded, dest: sp_global.bottomUpIn},
     {src: sp_global.bottomUpOut, dest: tm.bottomUpIn},      
     {src: encoder.bucket, dest: classifier.bucket},
     {src: tm.bottomUpOut, dest: classifier.pattern}]
    }}}
```
So, we will be calling addRegion( ) 4 times and Link( ) 4 times and then initialize.

**Comment by dkeeney** [comment]
Now, lets say we have a REST client that wants to run something like the napi_sine.cpp example.  Lets assume there is a server.exe built from code I will have in the examples.  It is started with the port number it will be listening on.  The REST client would need to configure itself with the ip address and port of the server we just started.  

The client then sends its first message to configure a Network object on the server.  This message would be a POST message passing our configuration string in JSON syntax and use a command "/configure".   If we were to use the curl command as the client it might look something like this:
```
curl -d '<our configuration string goes here>' -H "Content-Type: application/json" -X POST http://localhost:3000/configure
```
The response would either be an error or it would return an OK status and a data value which is the handle of the created and configured Network class.

**Comment by dkeeney** [comment]
Now we want to run this with some data.  We need to pass it two values for each iteration.  The first is the handle of the Network class and the second is the data to be passed to the encoder (along with some type indicator).  In our example, we are passing points on the sine wave.  So in a loop we iterate the sine wave.

Again with a command line curl it might be something like:
```
curl -k -i -X GET -H "Content-Type:application/json" http://localhost:3000/data/handle/<handle>/type/float/value/<value>
```
the value is our current sine wave point being passed in and the response (the predicted value) is returned.  Alternatively would could pass the handle, type and value in a POST rather than a GET request.  Depending on the type of encoder you might want to pass an array of data.  There are a lot of options that would be implemented depending on the application.

I hope this is enough to get the idea.

**Comment by dkeeney** [comment]
Rather than curl, I would expect a real application to use one of the REST client packages to send and receive messages.

**Comment by dkeeney** [comment]
As I start to document this I realize that perhaps I don't have the resources defined quite right.
I think perhaps the hierarchy of resource should be something like this:
```
Network
     Regions
          Parameters
          Inputs
          Outputs
     Links
```

So the RISK key structure would be something like...
```
/network/<id>/region/<region name>/param/<param name>
/network/<id>/region/<region name>/Input/<param name>
/network/<id>/region/<region name>/output/<param name>
/network/<id>/link/<link name>
```

So, I think I should modify the server to match this.

**Comment by breznak** [comment]
@dkeeney sorry it takes me that long to review, I'm keeping an eye on the mails and PR but I didn't have time to do a proper review yet. Promise to do it during Tue, Wed.

**Comment by dkeeney** [comment]
No problem.  I want to add more documentation anyway.

**Comment by breznak** [comment]
Fixed on #778 :champagne:
